common/ferrostar/src/navigation_controller/mod.rs

```diff
@@ -368,6 +373,55 @@ impl NavigationController {
     }
 }
 
+/// Functionality for the navigation controller that is exported. (or will be exported)
+impl NavigationRecording {
+    /// Creates a new navigation recorder with route configuration and initial state.
+    pub fn new(
+        route_config: NavigationControllerConfig,
+        initial_state: InitialNavigationState,
+    ) -> Self {
+        Self {
+            version: env!("CARGO_PKG_VERSION").to_string(),
+            initial_timestamp: Utc::now().timestamp(),
+            route_config,
+            initial_state,
+            events: Vec::new(),
+        }
+    }
+
+    /// Records a location update from the user during navigation.
+    pub fn record_location_update(&mut self, user_location: UserLocation) {
+        self.add_event(NavigationRecordingEventData::LocationUpdate { user_location });
+    }
+
+    /// Records a trip state update during navigation.
+    pub fn record_trip_state_update(&mut self, trip_state: TripState) {
+        self.add_event(NavigationRecordingEventData::TripStateUpdate { trip_state });
+    }
+
+    /// Records a route update during navigation.
+    pub fn record_route_update(&mut self, route: Route) {
+        self.add_event(NavigationRecordingEventData::RouteUpdate { route });
+    }
+
+    /// Records an error that occurred during navigation.
+    pub fn record_navigation_error(&mut self, error_message: String) {
+        self.add_event(NavigationRecordingEventData::Error { error_message });
+    }
```

Comment:
This will _probably_ break once you try exporting it. UniFFI doesn't support mutable self references in exports, because there is no way to guarantee that foreign code (e.g. Swift, Kotlin, etc.) will uphold the assumptions Rust makes. That's why we'll need to either:

1. Work out an efficiently appendable data  structure where every call consumes the original value and returns a new one with each call (like we do for the rest of the pure core already), OR
2. Introduce some sort of interior mutability (obviously not ideal).